source/extensions/common/dynamic_forward_proxy/dns_cache_impl.cc

```diff
@@ -224,82 +246,75 @@ void DnsCacheImpl::finishResolve(const std::string& host,
+  auto current_address = primary_host_info->host_info_->address();
+  if (new_address != nullptr && (current_address == nullptr || *current_address != *new_address)) {
     ENVOY_LOG(debug, "host '{}' address has changed", host);
-    primary_host_info.host_info_->address_ = new_address;
-    runAddUpdateCallbacks(host, primary_host_info.host_info_);
+    primary_host_info->host_info_->setAddress(new_address);
+    runAddUpdateCallbacks(host, primary_host_info->host_info_);
     address_changed = true;
     stats_.host_address_changed_.inc();
   }
 
   if (first_resolve || address_changed) {
-    updateTlsHostsMap();
+    primary_host_info->host_info_->setFirstResolveComplete();
+    notifyThreads(host);
   }
 
   // Kick off the refresh timer.
   // TODO(mattklein123): Consider jitter here. It may not be necessary since the initial host
   // is populated dynamically.
   if (status == Network::DnsResolver::ResolutionStatus::Success) {
     failure_backoff_strategy_->reset();
-    primary_host_info.refresh_timer_->enableTimer(refresh_interval_);
+    primary_host_info->refresh_timer_->enableTimer(refresh_interval_);
     ENVOY_LOG(debug, "DNS refresh rate reset for host '{}', refresh rate {} ms", host,
               refresh_interval_.count());
   } else {
     const uint64_t refresh_interval = failure_backoff_strategy_->nextBackOffMs();
-    primary_host_info.refresh_timer_->enableTimer(std::chrono::milliseconds(refresh_interval));
+    primary_host_info->refresh_timer_->enableTimer(std::chrono::milliseconds(refresh_interval));
     ENVOY_LOG(debug, "DNS refresh rate reset for host '{}', (failure) refresh rate {} ms", host,
               refresh_interval);
   }
 }
 
 void DnsCacheImpl::runAddUpdateCallbacks(const std::string& host,
                                          const DnsHostInfoSharedPtr& host_info) {
-  for (auto callbacks : update_callbacks_) {
+  for (auto* callbacks : update_callbacks_) {
     callbacks->callbacks_.onDnsHostAddOrUpdate(host, host_info);
   }
 }
 
 void DnsCacheImpl::runRemoveCallbacks(const std::string& host) {
-  for (auto callbacks : update_callbacks_) {
+  for (auto* callbacks : update_callbacks_) {
     callbacks->callbacks_.onDnsHostRemove(host);
   }
 }
 
-void DnsCacheImpl::updateTlsHostsMap() {
-  TlsHostMapSharedPtr new_host_map = std::make_shared<TlsHostMap>();
-  for (const auto& primary_host : primary_hosts_) {
-    // Do not include hosts that have not resolved at least once.
-    if (primary_host.second->host_info_->first_resolve_complete_) {
-      new_host_map->emplace(primary_host.first, primary_host.second->host_info_);
-    }
-  }
-
-  tls_slot_.runOnAllThreads([new_host_map](OptRef<ThreadLocalHostInfo> local_host_info) {
-    local_host_info->updateHostMap(new_host_map);
+void DnsCacheImpl::notifyThreads(const std::string& host) {
+  auto host_ptr = std::make_shared<const std::string>(host);
+  tls_slot_.runOnAllThreads([host_ptr](OptRef<ThreadLocalHostInfo> local_host_info) {
```

Comment:
I think you can just capture by copy here?

Reply:
Ah I missed this comment in my last update. A capture by copy here will make a string copy for each worker thread. Is that an acceptable cost?

Reply:
My guess is that in most cases it will be small object inlined into the lambda, but I'm not sure. It doesn't matter that much so whatever you think is best.

Reply:
Ok I'll leave it as a shared_ptr for now.